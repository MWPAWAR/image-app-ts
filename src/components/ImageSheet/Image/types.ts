import { ImageTypes } from '../../types'

export interface ImagePropsType {
  selectedImage: ImageTypes
  activeTab: string
  toggleDrawer: () => void
}
