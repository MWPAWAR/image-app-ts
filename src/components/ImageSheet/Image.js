import { useDispatch } from 'react-redux';
import imageStyles from './Image.module.css';
import classNames from 'classnames';
import { getMbsTextByBytes } from '../../utils/get-mbs-by-bytes';
import { toggleFavoriteImg } from '../../store/images/images';
import { resetSelectedImg } from '../../store/imageUiState/imageUiState';

const Image = ({ activeTab, selectedImage }) => {
  const dispatch = useDispatch();
  const handleFavtClick = () => {
    dispatch(toggleFavoriteImg(selectedImage.id));
    // This is to reset the selected img if the active tab is favorites.
    if (activeTab === 'favorites') dispatch(resetSelectedImg());
  };

  return (
    <div>
      <img
        src={selectedImage.url}
        className={imageStyles.image}
        alt={selectedImage.filename}
      />
      <div className={imageStyles.imgNameAndFavtIconContainer}>
        <div className={imageStyles.imageName}>{selectedImage.filename}</div>
        <div className={classNames(
          imageStyles.favtIcon,
          selectedImage.favorited ? imageStyles.favtIconFilled : imageStyles.favtIconOutlined
        )} onClick={handleFavtClick} />
      </div>
      <div className={imageStyles.imageSize}>{getMbsTextByBytes(selectedImage.sizeInBytes)}</div>
    </div>
  );
};

export default Image;
