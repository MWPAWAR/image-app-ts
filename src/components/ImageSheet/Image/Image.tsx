import classNames from 'classnames'
import imageStyles from './Image.module.css'
import { useAppDispatch } from '../../customHooks/hooks'
import { getMbsTextByBytes } from '../../../utils/get-mbs-by-bytes'
import { toggleFavoriteImg } from '../../../store/images/images'
import { resetSelectedImg } from '../../../store/imageUiState/imageUiState'
import { ImagePropsType } from './types'

const Image = ({ activeTab, selectedImage, toggleDrawer }: ImagePropsType) => {
  const dispatch = useAppDispatch()
  const handleFavtClick = (): void => {
    dispatch(toggleFavoriteImg({ id: selectedImage.id }))
    // Do this to reset the selected img if the active tab is favorited.
    if (activeTab === 'favorited') {
      toggleDrawer()
      dispatch(resetSelectedImg())
    }
  }

  return (
    <div>
      <img
        src={selectedImage.url}
        className={imageStyles.image}
        alt={selectedImage.filename}
      />
      <div className={imageStyles.imgNameAndFavtIconContainer}>
        <div className={imageStyles.imageName}>{selectedImage.filename}</div>
        <div
          data-testid="fav-icon"
          className={classNames(
            imageStyles.favtIcon,
            selectedImage.favorited
              ? imageStyles.favtIconFilled
              : imageStyles.favtIconOutlined
          )}
          onClick={handleFavtClick}
        />
      </div>
      <div className={imageStyles.imageSize}>
        {getMbsTextByBytes(selectedImage.sizeInBytes)}
      </div>
    </div>
  )
}

export default Image
