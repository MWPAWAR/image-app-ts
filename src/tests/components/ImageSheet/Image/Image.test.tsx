import { MemoryRouter } from 'react-router-dom'
import { Provider } from 'react-redux'
import renderer from 'react-test-renderer'
import configureStore from 'redux-mock-store'
import { render } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import Image from '../../../../components/ImageSheet/Image/Image'
import { imageMock1 } from '../../../mocks/image'
import { toggleDrawer } from '../../../test-utils'

describe('Image', () => {
  const mockStore = configureStore([])
  const store = mockStore({})
  jest.spyOn(store, 'dispatch')

  const ImageComponent = () => (
    <MemoryRouter>
      <Provider store={store}>
        <Image
          activeTab="favorited"
          selectedImage={imageMock1}
          toggleDrawer={toggleDrawer}
        />
      </Provider>
    </MemoryRouter>
  )

  it('Matches snapshot', () => {
    const tree = renderer.create(<ImageComponent />)
    expect(tree).toMatchSnapshot()
  })

  it('Calls dispatch if user clicks on fav icon', async () => {
    const { getByTestId } = render(<ImageComponent />)
    await userEvent.click(getByTestId('fav-icon'))
    expect(store.dispatch).toBeCalledTimes(2)
    expect(store.dispatch).toHaveBeenCalledWith({
      type: 'images/toggleFavoriteImg',
      payload: { id: imageMock1.id },
    })
    expect(store.dispatch).toHaveBeenCalledWith({
      type: 'imageUiState/resetSelectedImg',
    })
  })
})
