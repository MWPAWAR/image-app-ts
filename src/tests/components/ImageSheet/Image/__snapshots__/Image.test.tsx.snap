// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image Matches snapshot 1`] = `
<div>
  <img
    alt="tennessee_female_rubber.jpg"
    className="image"
    src="https://agencyanalytics-api.vercel.app/images/0.jpg"
  />
  <div
    className="imgNameAndFavtIconContainer"
  >
    <div
      className="imageName"
    >
      tennessee_female_rubber.jpg
    </div>
    <div
      className="favtIcon favtIconFilled"
      data-testid="fav-icon"
      onClick={[Function]}
    />
  </div>
  <div
    className="imageSize"
  >
    4.6 MB
  </div>
</div>
`;
